Loan writeback guide (#789)
codatio/codat-docs: docs/lending/guides/general-loan-writeback.md

sidebar_label: "Loan writeback"
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> ## What is it?

Reviewer: ```suggestion
## What is loan writeback?
```

> ## What is it?

Reviewer: ```suggestion
## Context

### What is it?
```

Would be good structurally to nest this context-setting intro bits under a header (even if just to make the table of contents better)

Loan writeback (also known as lending writeback) is the process of continuously updating an accounting platform with information on a loan. It helps maintain an accurate position of the loan during the entire lending cycle by recording the loan liability, any interest, fees, or repayments, and facilitating the reconciliation of bank transactions. 

> This guide covers the loan writeback procedure for general lending, such as term loans, but does not cover requirements specific to invoice finance.

Reviewer: This guide covers how to implement...?

Some language being specific about what the user will do and achieve with the guide?

:::info Mandatory loan writeback

Certain accounting platforms **require** lenders to continuously update their books with money lent to SMBs. For example, **Xero** obligates lenders going through the [App Partner certification](/integrations/accounting/xero/xero-app-partner-program) process to handle the writeback process. 

:::

> ## Why use it?

Reviewer: ```suggestion
### Why use it?
```

A bookkeeper can account for a loan in numerous ways in an accounting platform. For example, some bookkeepers may erroneously register a loan as a direct income or even a sales invoice. This results in loans being improperly recorded as revenue and repayments as operating costs. At the end of the reporting period, this can make it hard for the bookkeeper to close their books. 

By implementing loan writeback functionality in your application, you can make sure loan bookeeping is done regularly, correctly, and quickly, and always see an up-to-date state of the borrower's accounts.

> ## What's the process?

Reviewer: ```suggestion
### What's the process?
```

## Prerequisites

> * Check that you have [created a Codat company](/configure/portal/companies#add-a-new-company) that represents your SMB customer and linked it to an accounting platform. If you are using Codat for lending, it's likely you have already done this. You can also create and connect a test company to use while you build your solution.

Reviewer: ```suggestion
* Check that you have [created a Codat company](/configure/portal/companies#add-a-new-company) that represents your SMB customer and linked it to an accounting platform. If you are already using Codat for lending, it's likely you have already done this. You can also create and connect a test company to use while you build your solution.
```

Reviewer: We would suggest you don't test this on real customers anyway, so maybe pull out that last sentence and say 'you should' rather than can

    backend ->> codat: Create record (data type)
    codat -->> backend: Push operation key

    codat -->> backend: Push operation status webhook

    alt Status is successful
        backend ->> codat: Get push operation
        codat -->> backend: push operation
    end
```

> * If you are implementing loan writeback for Xero, make sure that Xero have enabled the *Xero Bank Feeds API* for your registered app. In order to create and update bank transactions, your application must be approved by [Xero's certification program](https://developer.xero.com/documentation/xero-app-store/app-partner-guides/app-partner-steps/).

Reviewer: Is there a way of testing/building pre-approval?

### Bank account

Loan writeback process operates with two bank accounts: a borrower's business bank account, where the money lent is deposited, and a lender's bank account, which is a virtual bank account in the accounting platform that acts as a container for lender transactions. 

First, your customer needs to choose one of their existing bank accounts. This account will be used to depost the loan. Call our [List bank accounts](/lending-api#/operations/list-accounting-bank-accounts) endpoint to retrieve the customer's existing bank accounts. 

<Tabs>
<TabItem value="nodejs" label="TypeScript">

```javascript
codatLending.accountingBankData.accounts.list({
    companyId: "8a210b68-6988-11ed-a1eb-0242ac120002",
    connectionId: "2e9d2c44-f675-40ba-8049-353bfcb5e171"
}).then((res: ListAccountingBankAccountsResponse) => {
if (res.statusCode == 200) {

>     // handle response

Reviewer: @dcoplowe any suggested way of handling? (Can always add this later)

</TabItem>

<TabItem value="python" label="Python">

```python
bank_accounts_list_request = operations.ListAccountingBankAccountsRequest(
    company_id='8a210b68-6988-11ed-a1eb-0242ac120002',
    connection_id='2e9d2c44-f675-40ba-8049-353bfcb5e171'
)

bank_accounts_list_response = codat_lending.accounting_bank_data.accounts.list(bank_accounts_list_request)
```
</TabItem>

<TabItem value="csharp" label="C#">

```csharp
var bankAccountsListResponse = await codatLending.AccountingBankData.Accounts.ListAsync(new ListAccountingBankAccountsRequest() {
    CompanyId = "8a210b68-6988-11ed-a1eb-0242ac120002",
    ConnectionId = "2e9d2c44-f675-40ba-8049-353bfcb5e171"
});
```
</TabItem>

<TabItem value="go" label="Go">

```go
ctx := context.Background()
bankAccountsListResponse, err := codatLending.AccountingBankData.Accounts.List(
    ctx, 
    operations.ListAccountingBankAccountsRequest{
        CompanyID: "8a210b68-6988-11ed-a1eb-0242ac120002",
        ConnectionID: "2e9d2c44-f675-40ba-8049-353bfcb5e171",
    })
```
</TabItem>

<TabItem value="http" label="HTTP">

```http
GET https://api.codat.io/companies/{companyId}/connections/{connectionId}/data/bankAccounts
```
</TabItem>

</Tabs>

> Display the response to the customer and allow them to select the account. Store the returned `id` and use it as the borrower's account Id in future operations. 

Reviewer: ```suggestion
Display the response to the customer and allow them to select the account. Store the returned `id` and use it as the borrower's `accountId` in future operations. 
```

Next, you need to create a new bank account in the borrower's accounting system. This will be the lender's - your - virtual bank account, used to track transactions associated with your lending activity. 

> Use our [Get create/update bank account model](/lending-api#/operations/get-create-update-bankAccounts-model) to get the expected data for the bank account creation request payload. Then, use the [Create bank account](/lending-api#/operations/create-bank-account) endpoint to create the new account in the accounting platform.

Reviewer: ```suggestion
To create a bank account:
1. Use our [Get create/update bank account model](/lending-api#/operations/get-create-update-bankAccounts-model) to get the expected data for the bank account creation request payload. The data required can vary depending on the platform.
2. Use the [Create bank account](/lending-api#/operations/create-bank-account) endpoint to create the new account in the accounting platform.
```

});
```
</TabItem>

<TabItem value="go" label="Go">

```go
ctx := context.Background()
createBankAccountResponse, err := codatLending.LoanWriteback.BankAccounts.Create(ctx, operations.CreateBankAccountRequest{
    AccountingBankAccount: &shared.AccountingBankAccount{
        AccountName: lending.String("saepe"),
        AccountNumber: lending.String("fuga"),
        AccountType: shared.AccountingBankAccountTypeCredit.ToPointer(),
        AvailableBalance: types.MustNewDecimalFromString("3595.08"),
        Balance: types.MustNewDecimalFromString("6130.64"),
        Currency: lending.String("USD"),
        IBan: lending.String("saepe"),
        Institution: lending.String("commodi"),
        NominalCode: lending.String("molestiae"),
        OverdraftLimit: types.MustNewDecimalFromString("2444.25"),
        SortCode: lending.String("error"),
    },
    CompanyID: "8a210b68-6988-11ed-a1eb-0242ac120002",
    ConnectionID: "2e9d2c44-f675-40ba-8049-353bfcb5e171",
})
```
</TabItem>

<TabItem value="http" label="HTTP">

```http
POST https://api.codat.io/companies/{companyId}/connections/{connectionId}/push/bankAccounts
```
</TabItem>

</Tabs>

> In response, you will receive account creation details which you can display to your customer. Similarly, store the `id` and use it in future transactions. 

Reviewer: ```suggestion
In response, you will receive account creation details which you can display to your customer. Just as you did with the `accountId`, store this `id` and use it in future transactions as.... 
```

What could they refer to this as? `lenderBankAccountId`?

### Supplier

> In order to create a spend money transaction, Codat requires suppler details that represent you, the lender, in your SMB's accounting system. 

Reviewer: ```suggestion
In order to create a *spend money* transaction, Codat requires suppler details that represent you, the lender, in your SMB's accounting system. 
```

Reviewer: 'In an accounting system, a lender would be represented as a Supplier.'

Make link to data type, or glossary

    "status": "Active",
    "type": "Asset"
}
```
</TabItem>

</Tabs>

In response, you will receive account creation details which you can display to your customer. Similarly, store the `id` and use it in future transactions. 

## Deposit the loan

Once you receive the configuration information, you are ready to deposit funds into the borrower's bank account. This is also known as *loan drawdown*, and it is a two-step process.  

For each drawdown, [create a bank transaction](/lending/guides/general-loan-writeback#create-bank-transaction) depositing the amount into the lender's bank account to make the funds available for drawdown. Then, [create a transfer](/lending/guides/general-loan-writeback#create-transfer) from the lender's bank account to the borrower's bank account. 

```mermaid
    sequenceDiagram
        participant backend as Your application 
        participant codat as Codat

        backend ->> codat: Create bank transaction (deposit to lender's account)
        codat -->> backend: bank transaction
        
        backend ->> codat: Create transfer (loan drawdown, lender -> bank account)
        codat -->> backend:  transfer
```

### Create bank transaction

> Use our [Get create bank account transactions model](/lending-api#/operations/get-create-bank-transactions-model) endpoint first to determine the parameters required for transaction creation. Then, use the [Create bank account transactions](/lending-api#/operations/create-bank-transactions) to deposit the amount into the lender's bank account with at least the following information included:

Reviewer: Generally, if we're saying 'then', we might want to think about turning it into a numbered list
